### PR TITLE
feat: RCM for Sales Transaction

### LINKED ISSUE #1800: Should RCM Sales be classified as Nil Rated? Related Fixes for GST Treatment
- [ ] Similar to purchase reverse charge
- [ ] New Tax Type: Sales Reverse Change
- [ ] Create templates only where it's required, create accounts and Template on change of GST Settings
- [ ] Validate correct usage
- [ ] GST Balance Report
- [ ] e-Invoice with RCM
- [ ] Virtual field for total RCM payable by Customer, Update relevant description in Standard GST Print format

### PR DESCRIPTION
**Pull request by ljain112**
Closes: #1800

### No Patch to create templates

Templates may need to be created and updated in GST Settings Manually.

* No use of template if RCM in sales is not applicable
* RCM Accounts may already exist for other cases. There could be multiple variable factors like GST Rate, Tax Category etc.
* Still updating of this in Item Tax Template will need to be manual as it's used differently by different companies.

### CODE REVIEW

File: india_compliance/gst_india/utils/transaction_data.py [hunk 1]
```diff
@@ -360,22 +365,19 @@ def update_item_tax_details(self, item_details, item):
             for tax in GST_TAX_TYPES[:3]
         )
 
+        total_value = self.rounded(item_details.taxable_value)
```
Comment (vorasmit) [inline india_compliance/gst_india/utils/transaction_data.py:368]: .

File: india_compliance/gst_india/utils/transaction_data.py [hunk 2]
```diff
@@ -131,6 +131,11 @@ def update_transaction_tax_details(self):
 
         # Other Charges
         current_total = 0
+
+        if self.doc.doctype == "Sales Invoice" and self.doc.is_reverse_charge:
```
Comment (vorasmit) [inline india_compliance/gst_india/utils/transaction_data.py]: Verify for purchase and e-Waybill

File: india_compliance/gst_india/utils/transaction_data.py [hunk 3]
```diff
@@ -360,22 +365,19 @@ def update_item_tax_details(self, item_details, item):
...
 
+        total_value = self.rounded(item_details.taxable_value)
+
+        if not (self.doc.doctype == "Sales Invoice" and self.doc.is_reverse_charge):
+            total_value += sum(
+                self.rounded(item_details.get(f"{tax}_amount", 0))
+                for tax in GST_TAX_TYPES
+            )
+
+        total_value = abs(total_value)
```
Comment (vorasmit) [inline india_compliance/gst_india/utils/transaction_data.py]: abs(rounded())

File: india_compliance/patches/post_install/update_sales_rcm_accounts.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+        "GST Account", {"account_type": ["=", "Sales Reverse Charge"]}
+    ):
+        return
+
+    companies = frappe.get_all("Company", filters={"country": "India"}, pluck="name")
+    for company in companies:
+        # determinig default gst rate
+        output_igst_account = get_gst_accounts_by_type(company, "Output").igst_account
+
+        default_gst_rate = frappe.db.get_value(
```
Comment (vorasmit) [inline india_compliance/patches/post_install/update_sales_rcm_accounts.py]: or 18%

File: india_compliance/patches/post_install/update_sales_rcm_accounts.py [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+        gst_accounts,
+        existing_account_list,
+        gst_settings,
+        "Sales Reverse Charge",
+    )
+
+    gst_settings.save()
+
+
+def update_item_tax_template(company):
```
Comment (vorasmit) [inline india_compliance/patches/post_install/update_sales_rcm_accounts.py]: not needed. custom implementations possible.

File: india_compliance/patches/post_install/update_sales_rcm_accounts.py [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+    default_taxes = frappe.get_file_json(get_data_file_path("tax_defaults.json"))
+    sales_tax_templates = (
+        default_taxes.get("chart_of_accounts").get("*").get("sales_tax_templates")
+    )
+    sales_rcm_tax_templates = []
+    for template in sales_tax_templates:
+        if template.get("title") in SALES_RCM_TEMPLATES:
+            sales_rcm_tax_templates.append(template)
+
+    for template in sales_rcm_tax_templates:
```
Comment (vorasmit) [inline india_compliance/patches/post_install/update_sales_rcm_accounts.py]: for loop needed?

File: india_compliance/patches/post_install/update_sales_rcm_accounts.py [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+
+            rate = (
+                gst_rate
+                if row["account_head"]["tax_rate"] == 18
+                else flt(gst_rate / 2, 3)
+            )
+
+            row["account_head"]["tax_rate"] = rate
+
+        make_taxes_and_charges_template(
```
Comment (vorasmit) [inline india_compliance/patches/post_install/update_sales_rcm_accounts.py]: check if account exists, and continue.

File: india_compliance/gst_india/overrides/item_tax_template.py [hunk 8]
```diff
@@ -39,15 +40,32 @@ def validate_tax_rates(doc):
     if not intra_state_accounts and not inter_state_accounts:
         return
 
+    valid_accounts = inter_state_accounts + inter_state_accounts
```
Comment (vorasmit) [inline india_compliance/gst_india/overrides/item_tax_template.py]: inter + intra

File: india_compliance/gst_india/overrides/item_tax_template.py [hunk 9]
```diff
@@ -39,15 +40,32 @@ def validate_tax_rates(doc):
...
+    rcm_accounts = []
+
+    for account_type in ("Sales Reverse Charge", "Purchase Reverse Charge"):
```
Comment (vorasmit) [inline india_compliance/gst_india/overrides/item_tax_template.py]: Only for sales

File: india_compliance/gst_india/overrides/item_tax_template.py [hunk 10]
```diff
@@ -39,15 +40,32 @@ def validate_tax_rates(doc):
...
+                ).values()
+            )
+        )
+
     invalid_tax_rates = {}
     for row in doc.taxes:
-        # check intra state
-        if row.tax_type in intra_state_accounts and doc.gst_rate != row.tax_rate * 2:
-            invalid_tax_rates[row.idx] = doc.gst_rate / 2
+        if row.tax_type not in valid_accounts:
```
Comment (vorasmit) [inline india_compliance/gst_india/overrides/item_tax_template.py]: similar implementation in js

File: india_compliance/gst_india/overrides/transaction.py [hunk 11]
```diff
@@ -1335,6 +1360,19 @@ def onload(doc, method=None):
     set_gst_breakup(doc)
 
 
+def set_reverse_charge_amount(doc, method=None, print_settings=None):
```
Comment (vorasmit) [inline india_compliance/gst_india/overrides/transaction.py]: to-remove

**Comment by vorasmit** [comment]
- [x] Refactor Validate Accounts
- [ ] Consistent backend for is GST Applicable
- [ ] Test Cases
- [ ] Docs

File: india_compliance/patches/post_install/update_sales_rcm_accounts.py [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+
+        for template in TEMPLATE:
+            template["tax_category"] = tax_categories.get(template["tax_category"], "")
+
+            for row in template.get("taxes"):
+                taxes = row["account_head"]
+                update_account_name(gst_accounts, account_names, taxes)
+                update_tax_rate(gst_rate, taxes)
+
+            make_taxes_and_charges_template(
```
Comment (vorasmit) [inline india_compliance/patches/post_install/update_sales_rcm_accounts.py:113]: Issue:
This will create template as per account name as specified in our template. This could be different in GST Settings for Output accounts.

This should be independent patch and we should create taxes and charges using existing output accounts in GST Settings.

Comment (ljain112) [inline india_compliance/patches/post_install/update_sales_rcm_accounts.py]: Already handling

File: india_compliance/patches/post_install/update_sales_rcm_accounts.py [hunk 13]
```diff
@@ -0,0 +1,222 @@
...
+
+    rate = gst_rate if taxes["tax_rate"] == 18 else flt(gst_rate / 2, 3)
+
+    taxes["tax_rate"] = rate
+
+
+def get_tax_categories():
+    rcm_tax_categories = {}
+
+    tax_categories = frappe.get_all(
```
Comment (vorasmit) [inline india_compliance/patches/post_install/update_sales_rcm_accounts.py]: . first category, else create

File: india_compliance/patches/post_install/update_sales_rcm_accounts.py [hunk 14]
```diff
@@ -0,0 +1,222 @@
...
+
+TEMPLATE = [
+    {
+        "title": "Output GST RCM In-state",
+        "taxes": [
+            {
+                "account_head": {
+                    "account_name": "sgst_account",
+                    "tax_rate": 9.0,
+                    "account_type": "account_type",
```
Comment (vorasmit) [inline india_compliance/patches/post_install/update_sales_rcm_accounts.py]: .

File: india_compliance/patches/post_install/update_sales_rcm_accounts.py [hunk 15]
```diff
@@ -0,0 +1,222 @@
...
+            "Account",
+            {
+                "company": company,
+                "account_name": ("in", SALES_RCM_ACCOUNTS),
+            },
+        ):
+            continue
+
+        gst_accounts = get_gst_accounts_by_type(company, "Output")
+        account_names = get_account_names(gst_accounts)
```
Comment (vorasmit) [inline india_compliance/patches/post_install/update_sales_rcm_accounts.py]: account_name_map

File: india_compliance/patches/post_install/update_sales_rcm_accounts.py [hunk 16]
```diff
@@ -0,0 +1,222 @@
...
+            )
+
+        update_gst_settings(company)
+
+
+def get_default_gst_rate(gst_accounts):
+    gst_rate = (
+        frappe.db.get_value(
+            "Sales Taxes and Charges",
+            filters={"account_head": gst_accounts.igst_account},
```
Comment (vorasmit) [inline india_compliance/patches/post_install/update_sales_rcm_accounts.py]: parent: template

File: india_compliance/patches/post_install/update_sales_rcm_accounts.py [hunk 17]
```diff
@@ -0,0 +1,222 @@
...
+
+
+def update_account_name(gst_accounts, account_names, taxes):
+    template_account = taxes["account_name"]
+    if template_account in SALES_RCM_ACCOUNTS:
+        return
+
+    account_info = account_names.get(gst_accounts.get(template_account))
+
+    if not account_info:
```
Comment (vorasmit) [inline india_compliance/patches/post_install/update_sales_rcm_accounts.py]: handle this correctly

File: india_compliance/patches/post_install/update_sales_rcm_accounts.py [hunk 18]
```diff
@@ -0,0 +1,222 @@
...
+
+    taxes["account_name"] = account_info.account_name
+    taxes["root_type"] = account_info.root_type
+
+
+def update_tax_rate(gst_rate, taxes):
+    if gst_rate == 18:
+        return
+
+    rate = gst_rate if taxes["tax_rate"] == 18 else flt(gst_rate / 2, 3)
```
Comment (vorasmit) [inline india_compliance/patches/post_install/update_sales_rcm_accounts.py]: readable

File: india_compliance/patches/post_install/update_sales_rcm_accounts.py [hunk 19]
```diff
@@ -0,0 +1,222 @@
...
+                "company": company,
+                "account_name": (
+                    "in",
+                    SALES_RCM_ACCOUNTS,
+                ),
+            },
+            ["account_name", "name"],
+            as_list=1,
+        )
+    )
```
Comment (vorasmit) [inline india_compliance/patches/post_install/update_sales_rcm_accounts.py]: if len in not 3

File: india_compliance/gst_india/overrides/item_tax_template.py [hunk 20]
```diff
@@ -39,15 +40,25 @@ def validate_tax_rates(doc):
...
+            doc.gst_rate / 2 if row.tax_type in intra_state_accounts else doc.gst_rate
+        )
+
+        if rcm_accounts and row.tax_type in rcm_accounts:
+            gst_rate = gst_rate * -1
 
-        # check inter state
-        elif row.tax_type in inter_state_accounts and doc.gst_rate != row.tax_rate:
-            invalid_tax_rates[row.idx] = doc.gst_rate
+        if row.tax_rate != gst_rate:
```
Comment (vorasmit) [inline india_compliance/gst_india/overrides/item_tax_template.py]: abs compare

File: india_compliance/gst_india/client_scripts/item_tax_template.js [hunk 21]
```diff
@@ -93,3 +98,22 @@ async function get_gst_accounts(frm) {
...
+
+
+async function get_company_sales_rcm_accounts(frm) {
+    const company = frm.doc.company;
+
+    // cache company gst accounts
+    if (!frm._company_rcm_accounts?.[company]) {
+        frm._company_rcm_accounts = frm._company_rcm_accounts || {};
+        const { message } = await frappe.call({
+            method: "india_compliance.gst_india.overrides.transaction.get_valid_gst_accounts_by_type",
```
Comment (vorasmit) [inline india_compliance/gst_india/client_scripts/item_tax_template.js]: smarter suggestions for purchase

**Comment by rehandev26** [comment]
Getting an error while installing?? @vorasmit @sagarvora 

image

**Comment by sagarvora** [comment]
> Getting an error while installing?? @vorasmit @sagarvora

@rehandev26 This was fixed in a newer version. Sorry for the inconvenience.
